Fix issue with non-base-10 integer literals in enums

chrisjbillington opened the pull request:
I encountered the following traceback when running autopxd2:
```
Traceback (most recent call last):
  File "/usr/bin/autopxd", line 10, in <module>
    sys.exit(cli())
  File "/usr/lib/python3.7/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/usr/lib/python3.7/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/usr/lib/python3.7/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/lib/python3.7/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/usr/lib/python3.7/site-packages/autopxd/__init__.py", line 94, in cli
    outfile.write(translate(infile.read(), infile.name, extra_cpp_args))
  File "/usr/lib/python3.7/site-packages/autopxd/__init__.py", line 73, in translate
    p.visit(parse(code, extra_cpp_args=extra_cpp_args, whitelist=whitelist))
  File "/usr/lib/python3.7/site-packages/autopxd/writer.py", line 20, in visit
    rv = super(AutoPxd, self).visit(node)
  File "/usr/local/lib/python3.7/site-packages/pycparser/c_ast.py", line 158, in visit
    return visitor(node)
  File "/usr/local/lib/python3.7/site-packages/pycparser/c_ast.py", line 165, in generic_visit
    self.visit(c)
  File "/usr/lib/python3.7/site-packages/autopxd/writer.py", line 20, in visit
    rv = super(AutoPxd, self).visit(node)
  File "/usr/local/lib/python3.7/site-packages/pycparser/c_ast.py", line 158, in visit
    return visitor(node)
  File "/usr/lib/python3.7/site-packages/autopxd/writer.py", line 101, in visit_Decl
    decls = self.collect(node)
  File "/usr/lib/python3.7/site-packages/autopxd/writer.py", line 159, in collect
    self.generic_visit(node)
  File "/usr/local/lib/python3.7/site-packages/pycparser/c_ast.py", line 165, in generic_visit
    self.visit(c)
  File "/usr/lib/python3.7/site-packages/autopxd/writer.py", line 20, in visit
    rv = super(AutoPxd, self).visit(node)
  File "/usr/local/lib/python3.7/site-packages/pycparser/c_ast.py", line 158, in visit
    return visitor(node)
  File "/usr/lib/python3.7/site-packages/autopxd/writer.py", line 63, in visit_Enum
    value = int(item.value.value)
ValueError: invalid literal for int() with base 10: '0x20203843'
```

There is an enum in the header file being parsed with a value as a hex literal `0x20203843`, but the code in autopx2 passes that string to `int()` which assumes base 10.

I'm not sure what this part of the code is doing, and why it needs to convert the string representation of the int to a Python int (the result does not end up in the pxd file), but in any case in this PR I've changed it to be able to parse any C integer literal. This means using `int(value, base=0)` to tell it to use the base implied by the prefix of the literal, but also modifying the prefix to match Python syntax when the literal is in octal, since Python and C differ there.

Commit:
Fix issue with non-base-10 integer literals in enums

gabrieldemarmiesse commented:
Thanks for the PR! Could you add tests to the test/ directory for both cases? With a very simple header file. You can see how other tests are made. It should be easy enough and would ensure that this bug doesn't appear again.

chrisjbillington commented:
Sure, will do.

Commit:
Add a test for non-base-10 integer literals in enums.

Commit:
Preserve base of enum constants when used elsewhere.

Also handle type suffixes (LluU) on enum values.

Commit:
Fix Py2 issue.

chrisjbillington commented:
I now understand what the code is for - it is storing the enum values so that they may be used in array dimension definitions.

I've now changed it so that the base of the integer as entered is preserved if used in array dims. i.e if the integer is written as `0xabcdL`, then it will end up in the .pxd as `0xabcdL`. The suffixes are legal Cython, so the only change that gets made is about octals.

Updated the test to check this.
